test/e2e/support/utils.js

```diff
@@ -70,3 +70,39 @@ export async function newDesktopBrowserPage() {
 	global.page = await browser.newPage();
 	await page.setViewport( { width: 1000, height: 700 } );
 }
+
+/**
+ * Wait for all XHR request to finish before continuing.
+ *
+ * @param {number} minDelay minimum delay in milliseconds to wait for before returning
+ * @param {number} maxDelay maximum delay in milliseconds to wait for before returning
+ *
+ * @return {Promise}        Promise.
+ */
+export async function waitForRequests( minDelay = 1000, maxDelay = 20000 ) {
```

Comment:
This function is exactly the reason I loved Cypress. This was handled automatically. Can't find the equivalent in Puppeteer. This is trying to be as reliable as I can.

Reply:
While not intuitively named, I wonder if this is equivalent:

```js
page.waitForNavigation( { waitUntil: 'networkidle0' } ); 
```

https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagewaitfornavigationoptions

(There are various methods which have `networkidle0` as well)

Reply:
Alternatively, instead of waiting for requests, you could wait for the DOM to be in a state you expect once the request completes:

https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagewaitforselectorselector-options

Reply:
If anything, this highlights that Cypress gives us crutches to poor ways of expressing intent 😄

Reply:
> If anything, this highlights that Cypress gives us crutches to poor ways of expressing intent

I don't agree with this, the idea in Cypress is when you do an action (click, move, anything) you wait for all the side effects to happen before continuing. I think it's great in terms of e2e maintainability.

> page.waitForNavigation( { waitUntil: 'networkidle0' } ); 

This is not the same thing, this expects a page load and then waits for the network.

> Alternatively, instead of waiting for requests, you could wait for the DOM to be in a state you expect once the request completes:

Yes, but this is not generic enough for me, we'd have to recreate the same thing over and over again because the UI change is not the same and also, sometimes there's no UI change, just a request that you need to wait for before quitting a page for instance which will cut the request.

Reply:
>wait for all the side effects to happen

These aren't incidental side effects though. We should have very well-defined expectations of what it is we're waiting for.

Reply:
>This is not the same thing, this expects a page load and then waits for the network.

By its name, sure, though I'm curious if it works just as well in practice here.

Reply:
I tried it, it doesn't work

Reply:
Think as though you're a user activating a plugin. Is your intuition to continue to pressing "Delete" _once all network requests have been satisfied_?

Reply:
The user don't know if there are side effects or not, he doesn't care and he doesn't know if there async or sync operations, for me, we should always wait.

Reply:
I think we're on the same page there; they shouldn't know there are operations, but if there are and their existence blocks continuation (which I'm assuming by presence of "waitFor"), we should express tests in the same way that a user would expect to continue.

`waitForRequests` is like saying a user would wait for their network activity to stop to know when to continue.

Reply:
Understood but as I see it `waitForRequests` is a frameworky thing that's not in `puppeteer` by default and not something we need to write on our own in theory :)

Reply:
Now I don't know if we're debating anymore 😄 But this is what I mean when saying that it was a crutch provided by Cypress and made for poor tests. A better test is one which adheres to the user's expectations for stepping through a task. A best test (and a best _user experience_) is one for which no waiting occurs at all. So if anything, needing to make these explicit drives us toward a better experience overall.

Which is all to say: I don't think we should have `waitForRequests` or anything like it, as it's [smelly](https://en.wikipedia.org/wiki/Code_smell).

Reply:
I'm fine updating the PR to use selectors but I still disagree with you. Why `waitForNavigation` exist in the first place, you could also replace it with waiting for UI

Reply:
What do you mean "waiting for UI" ? I am fine (enough) with methods like `waitForSelector`.

Reply:
I was thinking: what's the difference between . `page.waitForNavivation` and `waitForRequests` there's none IMO, both are helpers to avoid the hassle of testing Async stuff.

Why would we be fine with `waitForNavigation` and not `waitForRequests` since both could be replaced with `waitForSelector`

Reply:
It depends in what context you mean to use `waitForNavigation`.

If you mean for here, to wait for XHR requests to complete, then I would agree, it's no different and is not a good tool to use.

If you mean to its intended behavior to wait for a page to load, then I think it's slightly different since for a full page load, the user does use more of a judgment of network activity and page readiness to gauge when to continue (screen turns white, then loads). None of these are perfect, sure, and I suppose this is why Puppeteer provides options for `waitUntil`, including `load`, `domcontentloaded`, `networkidle0`, and `networkidle2`.

But I do agree it's only slightly different. In reality, while page load process plays some role, in the end the user ultimately waits for the thing they expect to see on the page (`waitForSelector`).

Reply:
I updated using `waitForSelector` let  me know what you think :)